Sources/ProcedureKit/PendingEvent.swift

```diff
@@ -0,0 +1,124 @@
+//
+//  ProcedureKit
+//
+//  Copyright © 2016 ProcedureKit. All rights reserved.
+//
+
+import Foundation
+import Dispatch
+
+/// PendingEvent encapsulates a reference to a future Procedure event, and can be used
+/// to ensure that asynchronous tasks are executed to completion *before* the future event.
+///
+/// While a reference to the PendingEvent exists, the event will not occur.
+///
+/// You cannot instantiate your own PendingEvent instances - only the framework
+/// itself creates and provides (in certain circumstances) PendingEvents.
+///
+/// A common use-case is when handling a WillExecute or WillFinish observer callback.
+/// ProcedureKit will provide your observer with a PendingExecuteEvent or a PendingFinishEvent.
+///
+/// If you must dispatch an asynchronous task from within your observer, but want to
+/// ensure that the observed Procedure does not execute/finish until your asynchronous task
+/// completes, you can use the Pending(Execute/Finish)Event like so:
+///
+/// ```swift
+/// procedure.addWillFinishObserver { procedure, errors, futureFinish in
+///     DispatchQueue.global().async {
+///         futureFinish.doBeforeEvent {
+///             // do something asynchronous
+///             // this block is guaranteed to complete before the procedure finishes
+///         }
+//      }
+/// }
+/// ```
+///
+/// Some of the built-in Procedure functions take an optional "before:" parameter,
+/// to which a PendingEvent can be directly passed. For example:
+///
+/// ```swift
+/// procedure.addWillFinishObserver { procedure, errors, futureFinish in
+///     // produce a new operation before the procedure finishes
+///     procedure.produce(BlockOperation { /* do something */ }, before: futureFinish)
+/// }
+/// ```
+///
+final public class PendingEvent: CustomStringConvertible {
+    public enum PendingEventType: CustomStringConvertible {
```

Comment:
`Type` is a bit annoying to use, per the Swift compiler:
```swift
Type member may not be named 'Type', since it would conflict with the 'foo.Type' expression
```

But perhaps just shorten it to `Event` instead?